mempool/reactor.go

```diff
@@ -21,20 +22,82 @@ const (
 	maxTxSize  = maxMsgSize - 8 // account for amino overhead of TxMessage
 
 	peerCatchupSleepIntervalMS = 100 // If peer is behind, sleep this amount
+
+	// UnknownPeerID is the peer ID to use when running CheckTx when there is
+	// no peer (e.g. RPC)
+	UnknownPeerID uint16 = 0
 )
 
 // MempoolReactor handles mempool tx broadcasting amongst peers.
+// It maintains a map from peer ID to counter, to prevent gossiping txs to the
+// peers you received it from.
 type MempoolReactor struct {
 	p2p.BaseReactor
 	config  *cfg.MempoolConfig
 	Mempool *Mempool
+	ids     *mempoolIDs
+}
+
+type mempoolIDs struct {
+	mtx       sync.RWMutex
+	peerMap   map[p2p.ID]uint16
+	nextID    uint16              // assumes that a node will never have over 65536 active peers
+	activeIDs map[uint16]struct{} // used to check if a given peerID key is used, the value doesn't matter
+}
+
+// Reserve searches for the next unused ID and assignes it to the peer.
+func (ids *mempoolIDs) ReserveForPeer(peer p2p.Peer) {
+	ids.mtx.Lock()
+	defer ids.mtx.Unlock()
+
+	curID := ids.nextPeerID()
+	ids.peerMap[peer.ID()] = curID
+	ids.activeIDs[curID] = struct{}{}
+}
+
+// nextPeerID returns the next unused peer ID to use.
+// This assumes that ids's mutex is already locked.
+func (ids *mempoolIDs) nextPeerID() uint16 {
+	_, idExists := ids.activeIDs[ids.nextID]
+	for idExists {
+		ids.nextID++
+		_, idExists = ids.activeIDs[ids.nextID]
+	}
+	curID := ids.nextID
+	ids.nextID++
+	return curID
+}
+
+// Reclaim returns the ID reserved for the peer back to unused pool.
+func (ids *mempoolIDs) Reclaim(peer p2p.Peer) {
```

Comment:
we should also remove removedID from all the transactions in the mempool!!!

Reply:
ok, because nextID is always increasing until it overflows the chance of that is very low. can be ignored for now

Reply:
```go 	
        // XXX (melekes): commented out for performance reasons
	//
	// Remove the sender if present to avoid a new node joining and not receiving
	// a tx because we think it already has it. For that two things should happen:
	//   1) node has to connect to over 65536 peers during its lifetime
	//   2) tx must reside in the mempool for the above duration
	//
	// for e := memR.Mempool.txs.Front(); e != nil; e = e.Next() {
	// 	memTx := e.Value.(*mempoolTx)
	// 	memTx.senders.Delete(id)
	// }
```

Reply:
Some context: The above code could  become necessary once 1) and 2) show to happen. It is unclear if this will ever be the case or not.